File: user-service/src/main/java/com/study/user/application/service/OAuth2Service.java
package com.study.user.application.service;

import com.study.user.application.dto.TokenResponse;
import com.study.user.domain.entity.AuthProvider;
import com.study.user.domain.entity.Role;
import com.study.user.domain.entity.User;
import com.study.user.domain.repository.UserRepository;
import com.study.user.infrastructure.security.JwtTokenProvider;
import com.study.user.infrastructure.security.TokenRedisService;
import lombok.RequiredArgsConstructor;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestClient;

import java.util.Map;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class OAuth2Service {

    private final UserRepository userRepository;
    private final JwtTokenProvider jwtTokenProvider;
    private final TokenRedisService tokenRedisService;
    @org.springframework.beans.factory.annotation.Qualifier("oAuth2RestClient")
    private final RestClient restClient;

    @Transactional
    public TokenResponse loginWithGoogle(String accessToken) {
        Map<String, Object> userInfo = restClient.get()
                .uri("https://www.googleapis.com/oauth2/v3/userinfo")
                .header("Authorization", "Bearer " + accessToken)
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});

        String email = (String) userInfo.get("email");
        String name = (String) userInfo.get("name");
        String providerId = (String) userInfo.get("sub");

        User user = findOrCreateOAuth2User(email, name, AuthProvider.GOOGLE, providerId);
        return issueAndStoreTokens(user);
    }

    @Transactional
    public TokenResponse loginWithKakao(String accessToken) {
        Map<String, Object> userInfo = restClient.get()
                .uri("https://kapi.kakao.com/v2/user/me")
                .header("Authorization", "Bearer " + accessToken)
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});

        Map<String, Object> kakaoAccount = (Map<String, Object>) userInfo.get("kakao_account");
        Map<String, Object> profile = (Map<String, Object>) kakaoAccount.get("profile");

        String email = (String) kakaoAccount.get("email");
        String name = (String) profile.get("nickname");
        String providerId = String.valueOf(userInfo.get("id"));

        User user = findOrCreateOAuth2User(email, name, AuthProvider.KAKAO, providerId);
        return issueAndStoreTokens(user);
    }

    private User findOrCreateOAuth2User(String email, String name, AuthProvider provider, String providerId) {
        Optional<User> existingUser = userRepository.findByEmail(email);

        if (existingUser.isPresent()) {
            User user = existingUser.get();
            user.updateOAuth2Info(name, providerId);
            return user;
        }

        User newUser = User.builder()
                .email(email)
                .name(name)
                .role(Role.USER)
                .provider(provider)
                .providerId(providerId)
                .build();
        return userRepository.save(newUser);
    }

    private TokenResponse issueAndStoreTokens(User user) {
        String accessToken = jwtTokenProvider.createAccessToken(user.getId(), user.getEmail(), user.getRole().name());
        String refreshToken = jwtTokenProvider.createRefreshToken(user.getId());
        tokenRedisService.saveRefreshToken(user.getId(), refreshToken, 604800000L);
        return new TokenResponse(accessToken, refreshToken);
    }
}

File: user-service/src/main/kotlin/com/study/user/application/service/OAuth2Service.kt
package com.study.user.application.service

import com.study.common.exception.InvalidRequestException
import com.study.user.application.dto.TokenResponse
import com.study.user.domain.entity.AuthProvider
import com.study.user.domain.entity.Role
import com.study.user.domain.entity.User
import com.study.user.domain.repository.UserRepository
import com.study.user.infrastructure.security.JwtTokenProvider
import com.study.user.infrastructure.security.TokenRedisService
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.core.ParameterizedTypeReference
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.client.RestClient

@Service
class OAuth2Service(
    private val userRepository: UserRepository,
    private val jwtTokenProvider: JwtTokenProvider,
    private val tokenRedisService: TokenRedisService,
    @Qualifier("oAuth2RestClient") private val restClient: RestClient
) {
    @Transactional
    fun loginWithGoogle(accessToken: String): TokenResponse {
        val userInfo = restClient.get()
            .uri("https://www.googleapis.com/oauth2/v3/userinfo")
            .header("Authorization", "Bearer $accessToken")
            .retrieve()
            .body(object : ParameterizedTypeReference<Map<String, Any>>() {})
            ?: throw InvalidRequestException("Failed to fetch user info from OAuth2 provider")

        val email = userInfo["email"] as String
        val name = userInfo["name"] as String
        val providerId = userInfo["sub"] as String

        val user = findOrCreateOAuth2User(email, name, AuthProvider.GOOGLE, providerId)
        return issueAndStoreTokens(user)
    }

    @Transactional
    @Suppress("UNCHECKED_CAST")
    fun loginWithKakao(accessToken: String): TokenResponse {
        val userInfo = restClient.get()
            .uri("https://kapi.kakao.com/v2/user/me")
            .header("Authorization", "Bearer $accessToken")
            .retrieve()
            .body(object : ParameterizedTypeReference<Map<String, Any>>() {})
            ?: throw InvalidRequestException("Failed to fetch user info from OAuth2 provider")

        val kakaoAccount = userInfo["kakao_account"] as Map<String, Any>
        val profile = kakaoAccount["profile"] as Map<String, Any>
        val email = kakaoAccount["email"] as String
        val name = profile["nickname"] as String
        val providerId = userInfo["id"].toString()

        val user = findOrCreateOAuth2User(email, name, AuthProvider.KAKAO, providerId)
        return issueAndStoreTokens(user)
    }

    private fun findOrCreateOAuth2User(
        email: String, name: String, provider: AuthProvider, providerId: String
    ): User {
        val existing = userRepository.findByEmail(email)
        return if (existing.isPresent) {
            existing.get().also { it.updateOAuth2Info(name, providerId) }
        } else {
            userRepository.save(User(
                email = email, name = name, role = Role.USER,
                provider = provider, providerId = providerId
            ))
        }
    }

    private fun issueAndStoreTokens(user: User): TokenResponse {
        val accessToken = jwtTokenProvider.createAccessToken(user.id, user.email, user.role.name)
        val refreshToken = jwtTokenProvider.createRefreshToken(user.id)
        tokenRedisService.saveRefreshToken(user.id, refreshToken, 604800000L)
        return TokenResponse(accessToken, refreshToken)
    }
}
